### PR TITLE
[UNLANDED] Create prototype for AST rewriter

### LINKED ISSUE #45449: [FX] Write an AST rewriter
https://github.com/pytorch/pytorch/pull/45188#issuecomment-698644988

### LINKED ISSUE #45681: [FX] Handling of asserts is unergonomic
To address https://github.com/pytorch/pytorch/issues/44849, https://github.com/pytorch/pytorch/pull/45188 added torch.Assert, which is a traceable op that can be used in place of assert. However, this requires users to explicitly replace `assert` python statements with a callsite to that op. This is not very ergonomic and ideally we would be able to handle assert statements programmatically.

Suggestions for implementation:
- https://github.com/pytorch/pytorch/pull/45188#pullrequestreview-495251084
- https://github.com/pytorch/pytorch/pull/45188#issuecomment-698644988

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46410 Create prototype for AST rewriter**

Differential Revision: [D24665950](https://our.internmc.facebook.com/intern/diff/D24665950)

### CODE REVIEW

File: torch/fx/experimental/test_rewriter.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import torch
```
Comment (jamesr66a) [inline torch/fx/experimental/test_rewriter.py:1]: Can you move this file to `test/` or move these test cases to `test/test_fx_experimental.py`?

File: torch/fx/experimental/rewriter.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        source_ast = ast.parse(normalized_str)
+        dest_ast = ast.fix_missing_locations(self.visit(source_ast))
+        dest_str = astor.to_source(dest_ast)
+
+        # Pull out the compiled fucntion from the newly-created Module
+        code = compile(dest_ast, "", "exec")
+        globals_dict = {}
+        globals_dict = copy.copy(fn.__globals__)
+        exec(code, globals_dict)
+        fn_compiled = globals_dict["forward"]
```
Comment (jamesr66a) [inline torch/fx/experimental/rewriter.py:41]: This works for `forward` on `Module`s, but I think this is going to break for a top-level traceable function that's not called `forward`. Can you add a test for that case? And here's something I think we can try:

```
        globals_dict = copy.copy(fn.__globals__)
        keys_before = set(globals_dict.keys())
        exec(code, globals_dict)
        new_keys = list(set(globals_dict.keys()) - keys_before)
        assert len(new_keys) == 1
        fn_compiled = globals_dict[new_keys[0]]
```

This will figure out the name(s) that were introduced into globals in the course of executing the code object. Since we're specifically executing a code object for a single object, I believe there should only be a single new key introduced here. Also this can be tested like:

```    def test_assert_free_fn(self):
        def foo(x):
            return torch.relu(x)

        traced = symbolic_trace(foo)
        # Add more checks here too
```

File: torch/fx/experimental/rewriter.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            call_node_args.append(node.msg)
+        else:
+            call_node_args.append(ast.Constant(value="", kind=None))
+        call_node = ast.Call(
+            func=ast.Attribute(value=ast.Name(id="torch", ctx=ast.Load()),
+                               attr="Assert", ctx=ast.Load()),
+            args=call_node_args,
+            keywords=[])
+
+        # Workaround for astor
```
Comment (jamesr66a) [inline torch/fx/experimental/rewriter.py:59]: I don't think this is so much a workaround for `astor` so much as making sure that the returned node conforms to the Python AST grammar

File: torch/fx/experimental/rewriter.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    To support swapping out an AST node, define a new `visit` method on
+    that node. For more details, see:
+    https://docs.python.org/3/library/ast.html#ast.NodeTransformer
+    '''
+
+    def rewrite(self, fn: FunctionType):
+
+        # Get the source string, align everything with the `def`, then
+        # remote unnecessary indentation
```
Comment (jamesr66a) [inline torch/fx/experimental/rewriter.py:22]: Can you add a before/after example to this comment showing what it does? It took me a little while to understand what exactly was going on

Comment (jamesr66a) [reply torch/fx/experimental/rewriter.py:22]: BTW this is actually also an open issue for the TorchScript AST-based frontend https://github.com/pytorch/pytorch/issues/25043 so you could also submit a patch there (i.e. frontend.py) if you're so inclined

File: torch/fx/experimental/rewriter.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        split_source = source_str.split('\n')
+        aligned_source = [whitespace_prefix + s for s in split_source[1:]]
+        aligned_source.insert(0, split_source[0])
+        aligned_str = '\n'.join(aligned_source)
+        normalized_str = textwrap.dedent(aligned_str)
+
+        # Rewrite the original AST
+        source_ast = ast.parse(normalized_str)
+        dest_ast = ast.fix_missing_locations(self.visit(source_ast))
+        dest_str = astor.to_source(dest_ast)
```
Comment (jamesr66a) [inline torch/fx/experimental/rewriter.py:34]: Please remove as discussed :)

File: torch/fx/experimental/rewriter.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        call_node_args = [node.test]
+        if node.msg:
+            call_node_args.append(node.msg)
+        else:
+            call_node_args.append(ast.Constant(value="", kind=None))
+        call_node = ast.Call(
+            func=ast.Attribute(value=ast.Name(id="torch", ctx=ast.Load()),
+                               attr="Assert", ctx=ast.Load()),
+            args=call_node_args,
+            keywords=[])
```
Comment (jamesr66a) [inline torch/fx/experimental/rewriter.py:57]: You can save a few lines by using `ast.parse` itself to generate nodes for fixed constructs rather than having to manually construct them, e.g.

```
    def visit_Assert(self, node):
        # Create the Call node
        call_node = ast.parse('torch.Assert()', mode='eval').body
        msg = node.msg if node.msg else ast.Constant(value="", kind=None)
        call_node.args = [node.test, msg]

        # Workaround for astor
        expr_wrapper = ast.Expr(value=call_node)
<...>
```

File: torch/fx/experimental/test_rewriter.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+class TestASTRewriter(JitTestCase):
+
+    def test_assert_no_msg(self):
+        class M(torch.nn.Module):
+            def forward(self, a, b):
+                assert a == b
+                return a + b
+        m = M()
+        traced = symbolic_trace(m)
+        traced.graph.lint(traced)
```
Comment (jamesr66a) [inline torch/fx/experimental/test_rewriter.py:16]: I think for all these tests you would want to do two additional things:
1. Check the IR to make sure there's a `call_function` node with `target == torch.Assert`
2. Ensure that when the code is run, the assert throws when it's supposed to and doesn't throw when it's not supposed to. You can use `self.assertRaisesRegex` to intercept assertions and verify that they are thrown

File: torch/fx/symbolic_trace.py [hunk 8]
```diff
@@ -162,7 +166,6 @@ def module_call_wrapper(mod, *args, **kwargs):
             else:
                 return _create_proxy(self, 'call_module', module_qualified_name, args, kwargs)
         try:
-            torch.nn.Module.__call__ = module_call_wrapper
```
Comment (jamesr66a) [inline torch/fx/symbolic_trace.py:165]: Why this change? This will break tracing through nested modules, I think

File: torch/fx/symbolic_trace.py [hunk 9]
```diff
@@ -119,7 +120,7 @@ def is_leaf_module(self, m: torch.nn.Module, module_qualified_name : str) -> boo
         """
         return m.__module__.startswith('torch.nn') and not isinstance(m, torch.nn.Sequential)
 
-    def trace(self, root: Union[torch.nn.Module, Callable]) -> Graph:
+    def trace(self, root: Union[torch.nn.Module, Callable], rewrite: bool = True) -> Graph:
```
Comment (jamesr66a) [inline torch/fx/symbolic_trace.py:123]: Instead of modifying `trace`, can we just expose a function in `rewriter.py`? I think it would look something like:

```
def rewrite(fn : Union[torch.nn.Module, Callable]) -> Union[torch.nn.Module, Callable]:
  if isinstance(fn, torch.nn.Module):
    # Rewrite this module's forward() and all of its recursive children's
    # forward. Return the new rewritten module hierarchy.
    def rewrite_module(m : torch.nn.Module):
      new_m = copy.copy(m)
      for name, child in new_m.named_children():
          new_m[name] = rewrite_module(child)
      return AST_Rewriter.rewrite(new_m)
    return rewrite_module(fn)
  else:
    # Rewrite this single free function
    return AST_Rewriter.rewrite(fn)
```

We could optionally provide a rewriting Tracer implementation as well:

```
class RewritingTracer(torch.fx.Tracer):
  def trace(self, root: Union[torch.nn.Module, Callable]) -> Graph:
    return super().trace(torch.fx.experimental.rewriter.rewrite(root))
```

I think this all should cleanly decouple this functionality from the Tracer itself

**Review by jamesr66a — COMMENTED** [review]
image

File: torch/fx/experimental/rewriter.py (hunk 4 above)
Comment (ansley) [reply torch/fx/experimental/rewriter.py:22]: Submitted! When my PR goes through, I can call the `normalize_source_lines` function from here, which will make the code shorter and cleaner.

File: torch/fx/symbolic_trace.py [hunk 10]
```diff
@@ -157,7 +174,7 @@ def proxy_placeholder(name: str):
                 args.append(proxy_placeholder('**' + next(names_iter)))
             root_fn = _patch_function(root_fn, len(args))
 
-        return root_fn, args
+        return fn_for_analysis, args
```
Comment (jamesr66a) [inline torch/fx/symbolic_trace.py:177]: This isn't correct: `fn_for_analysis` is just for analysis, if you return it here then we'll be tracing through a function that's been stripped of various things (wrappers, etc).

Can you please just revert the changes to this file and add `_rewrite_` to `rewriter.py`, and then call that within a `RewritingTracer` class also in that file?

Since this is super experimental I don't think we want the entire FX test surface to go through this code like it is right now

File: test/test_fx_experimental.py [hunk 11]
```diff
@@ -157,5 +158,98 @@ def forward(self, a, b, offset):
...
+                assert a == b
+                return a + b
+        m = M()
+        traced = symbolic_trace(m)
+
+        # Make sure the graph is well-formed
+        traced.graph.lint(traced)
+
+        # Check the IR to make sure there's a call_function node with target == "Assert"
+        self.assertTrue(any(node.op == 'call_function' and get_target_name(node.target) == "Assert" for node in traced.graph.nodes))
```
Comment (jamesr66a) [inline test/test_fx_experimental.py:174]: Instead of using `get_target_name` here and below, can you just compare `node.target == torch.Assert`?

File: test/test_fx_experimental.py [hunk 12]
```diff
@@ -157,5 +158,98 @@ def forward(self, a, b, offset):
...
+
+        # Ensure that the assert throws when it's supposed to and doesn't throw when it's not supposed to
+        traced(3, 3)
```
Comment (jamesr66a) [inline test/test_fx_experimental.py:177]: Probably worth it to also check that the output is still correct:
`self.assertEqual(traced(3, 3), m(3, 3))`

Here and below

File: torch/fx/experimental/rewriter.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        keys_before = set(globals_dict.keys())
+        exec(code, globals_dict)
+        new_keys = list(set(globals_dict.keys()) - keys_before)
+        assert len(new_keys) == 1
+        fn_compiled = globals_dict[new_keys[0]]
+
+        # Return the correct FunctionType object
+        return fn_compiled
+
+    def visit_Assert(self, node):
```
Comment (jamesr66a) [inline torch/fx/experimental/rewriter.py:43]: Would be nice to have a docstring for this function describing what transformation is happening at a high level

**Review by jamesr66a — COMMENTED** [review]
Cool! Just a couple more things

File: torch/fx/symbolic_trace.py [hunk 14]
```diff
@@ -37,6 +38,21 @@ def _patch_function(fn: FunctionType, nargs: int) -> FunctionType:
     # we can't call this function normally, otherwise it would try to unpack them
     # instead, let's make python think that args and kwargs are normal variables
 
+def _rewrite(fn : Union[torch.nn.Module, Callable]) -> Union[torch.nn.Module, Callable]:
+    if isinstance(fn, torch.nn.Module):
+        # Rewrite this module's forward() and all of its recursive children's
+        # forward. Return the new rewritten module hierarchy.
+        def rewrite_module(m : torch.nn.Module):
+            new_m = copy.copy(m)
```
Comment (jamesr66a) [inline torch/fx/symbolic_trace.py:46]: OK in light of the `type(m)` issue, I think that you'll have to make a wrapper instead of using `copy.copy`. So:

```
        def rewrite_module(m : torch.nn.Module):
            class RewrittenModule(torch.nn.Module):
                def __init__(self, orig):
                    super().__init__()
                    self.__dict__ = copy.copy(orig.__dict__)
            RewrittenModule.forward = AST_Rewriter().rewrite(m.forward)
            new_m = RewrittenModule(m)
            for name, child in new_m.named_children():
                new_m[name] = rewrite_module(child)
            return new_m
```

File: torch/fx/symbolic_trace.py (hunk 10 above)
Comment (jamesr66a) [reply torch/fx/symbolic_trace.py:160]: These changes are still here but they should be reverted

File: torch/fx/node.py [hunk 15]
```diff
@@ -20,6 +20,13 @@
     BaseArgumentTypes
 ]]
 
+
+def get_target_name(target: Target) -> str:
+    if callable(target):
+        return target.__name__
+    return target
```
Comment (jamesr66a) [inline torch/fx/node.py:27]: Doesn't seem like this is used anymore

File: test/test_fx_experimental.py [hunk 16]
```diff
@@ -157,5 +163,110 @@ def forward(self, a, b, offset):
...
+        with self.assertRaisesRegex(AssertionError, error_msg):
+            traced(3, 5)
+
+        # Confirm that the output is correct
+        self.assertEqual(traced(3, 3), m(3, 3))
+
+    def test_traceable_function_with_nonstandard_name(self):
+        def foo(x):
+            return torch.relu(x)
+        traced = symbolic_trace(foo)
```
Comment (jamesr66a) [inline test/test_fx_experimental.py:269]: Probably need `symbolic_trace_with_rewrite` here as well, right?

**Review by jamesr66a — APPROVED** [review]
Looks good, just fix up the small things

**Comment by ngimel** [comment]
Unlanding, breaks CI
```
Traceback (most recent call last):
  File "test_type_hints.py", line 217, in test_run_mypy
    self.fail(f"mypy failed: {stdout} {stderr}")
AssertionError: mypy failed: torch/fx/experimental/rewriter.py:53: error: "AST" has no attribute "body"  [attr-defined]
torch/fx/experimental/rewriter.py:79: error: Argument 1 to "rewrite" of "AST_Rewriter" has incompatible type "Callable[..., Any]"; expected "FunctionType"  [arg-type]
torch/fx/experimental/rewriter.py:82: error: "Tensor" not callable  [operator]
torch/fx/experimental/rewriter.py:87: error: Argument 1 to "rewrite" of "AST_Rewriter" has incompatible type "Callable[..., Any]"; expected "FunctionType"  [arg-type]
Found 4 errors in 1 file (checked 1127 source files)
 ```
